zingo-sync/src/scan.rs

```diff
@@ -524,3 +545,143 @@ fn collect_nullifiers(
+        }
+
+        // wallet block must exist, otherwise the transaction will not have access to essential data such as the time it was mined
+        if let Some(wallet_block) = wallet_blocks.get(&block_height) {
+            if !wallet_block.txids().contains(&transaction.txid()) {
+                panic!("txid is not found in the wallet block at the transaction height!");
+            }
+        } else {
+            panic!("wallet block at transaction height not found!");
+        }
+
+        let wallet_transaction = scan_transaction(
+            parameters,
+            scanning_keys,
+            transaction,
+            block_height,
+            &decrypted_note_data,
+        )
+        .unwrap();
+        wallet_transactions.insert(txid, wallet_transaction);
+    }
+
+    Ok(wallet_transactions)
+}
+
+fn scan_transaction<P>(
+    parameters: &P,
+    scanning_keys: &ScanningKeys,
+    transaction: Transaction,
+    block_height: BlockHeight,
+    decrypted_note_data: &DecryptedNoteData,
+) -> Result<WalletTransaction, ()>
+where
+    P: Parameters,
+{
+    // TODO: price?
+    let zip212_enforcement = zip212_enforcement(parameters, block_height);
+    let mut sapling_notes: Vec<SaplingNote> = Vec::new();
+    let mut orchard_notes: Vec<OrchardNote> = Vec::new();
+
+    if let Some(bundle) = transaction.sapling_bundle() {
+        let sapling_keys: Vec<sapling_crypto::keys::PreparedIncomingViewingKey> = scanning_keys
+            .sapling()
+            .iter()
+            .map(|(_, key)| key.prepare())
+            .collect();
+        let sapling_outputs: Vec<(SaplingDomain, OutputDescription<GrothProofBytes>)> = bundle
+            .shielded_outputs()
+            .iter()
+            .map(|output| (SaplingDomain::new(zip212_enforcement), output.clone()))
+            .collect();
+
+        scan_notes::<SaplingDomain, OutputDescription<GrothProofBytes>, SaplingNote>(
+            &mut sapling_notes,
+            transaction.txid(),
+            &sapling_keys,
+            &sapling_outputs,
+            &decrypted_note_data.sapling_nullifiers_and_positions,
+        )
+        .unwrap();
+    }
+
+    if let Some(bundle) = transaction.orchard_bundle() {
+        let orchard_keys: Vec<orchard::keys::PreparedIncomingViewingKey> = scanning_keys
+            .orchard()
+            .iter()
+            .map(|(_, key)| key.prepare())
+            .collect();
+        let orchard_actions: Vec<(OrchardDomain, Action<Signature<SpendAuth>>)> = bundle
+            .actions()
+            .iter()
+            .map(|action| (OrchardDomain::for_action(action), action.clone()))
+            .collect();
+
+        scan_notes::<OrchardDomain, Action<Signature<SpendAuth>>, OrchardNote>(
+            &mut orchard_notes,
+            transaction.txid(),
+            &orchard_keys,
+            &orchard_actions,
+            &decrypted_note_data.orchard_nullifiers_and_positions,
+        )
+        .unwrap();
+    }
+
```

Comment:
// TODO: scan transparent?

Reply:
indeed! i have added a TODO to my latest PR